### PR TITLE
fix: fix the issue that duplicated CR metrics are generated when CRD is deleted/applied multiple times

### PR DESCRIPTION
**Pull request by CatherineF-dev**
**What this PR does / why we need it**: https://github.com/kubernetes/kube-state-metrics/issues/2223

```
# It returns same metrics data many times when CRD is deleted/applied multiple times
 curl localhost:8089/metrics
# HELP cr_creationtimestamp
# TYPE cr_creationtimestamp gauge
cr_creationtimestamp{customresource_group="example.com",customresource_kind="Bar",customresource_version="v1",name="mybar"} 1.701828773e+09
# HELP cr_resourceversion
# TYPE cr_resourceversion gauge
cr_resourceversion{customresource_group="example.com",customresource_kind="Bar",customresource_version="v1",name="mybar"} 909919
# HELP cr_creationtimestamp
# TYPE cr_creationtimestamp gauge
cr_creationtimestamp{customresource_group="example.com",customresource_kind="Bar",customresource_version="v1",name="mybar"} 1.701828773e+09
# HELP cr_resourceversion
# TYPE cr_resourceversion gauge
cr_resourceversion{customresource_group="example.com",customresource_kind="Bar",customresource_version="v1",name="mybar"} 909919
```

**How does this change affect the cardinality of KSM**: *(increases, decreases or does not change cardinality)*

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Mitigation for https://github.com/kubernetes/kube-state-metrics/issues/2223

### CODE REVIEW

File: internal/store/builder.go [hunk 1]
```diff
@@ -119,9 +119,22 @@ func (b *Builder) WithEnabledResources(r []string) error {
 	sort.Strings(sortedResources)
 
 	b.enabledResources = append(b.enabledResources, sortedResources...)
+	b.enabledResources = unique(b.enabledResources)
```
Comment (mrueg) [inline internal/store/builder.go:122]: Does it make sense to have a map[string] bool here?

**Comment by dgrisonnet** [comment]
I don't think this will fix the inherent issue. It seems that the mechanism detecting the deletion of CRD matched by the ksm config is not working properly. Normally when that happens, the internal store should be deleted alongside its informer because we don't want to expose metrics about this resource anymore.

**Comment by dgrisonnet** [comment]
To be clearer, the scenario where a specific resource appears twice in ksm code should never happen, so we should understand why it happened and fix it so that it appears only once and is correctly cleaned up when needed.

**Comment by CatherineF-dev** [comment]
got it, agree. Will debug again to find root cause.

**Comment by CatherineF-dev** [comment]
It should be related to https://github.com/kubernetes/kube-state-metrics/pull/1851 where adding PollForCacheUpdates function.

**Comment by CatherineF-dev** [comment]
It didn't delete removed CR https://github.com/rexagod/kube-state-metrics/blob/25a1d8da057cf761d614c59a52785335d34082d1/internal/discovery/discovery.go#L232.


How about creating a new issue and merging this PR as a mitigation?
I think it needs some changes around https://github.com/kubernetes/kube-state-metrics/pull/1851.

cc @dgrisonnet

File: internal/store/builder.go (hunk 1 above)
Comment (CatherineF-dev) [reply internal/store/builder.go:122]: I think using unique is fine as a mitigation.

It didn't remove deleted CR https://github.com/rexagod/kube-state-metrics/blob/25a1d8da057cf761d614c59a52785335d34082d1/internal/discovery/discovery.go#L232. After this is fixed, we can remove unique.

**Review by logicalhan — COMMENTED** [review]
/triage accepted
/assign @dgrisonnet

**Comment by korjek** [comment]
any updates here?
It leads to loosing the metrics, which is not good.

**Pull request closed without merging**